### PR TITLE
msg/async/rdma: refactor rx buffer pool allocator

### PR DESCRIPTION
**Pull request by alex-mikheev**
@yuyuyu101 please take a look

Pool allocator now has a context. Each context can have a
different configuration. It allows to use buffer pool safely
when there are multiple RDMAStack running in parrallel.

Do not fail on assertion when out of memory for rx buffers.
Instead log warning and try to recover.

Signed-off-by: Alex Mikheev <alexm@mellanox.com>

Conflicts:
	src/msg/async/rdma/Infiniband.h

### CODE REVIEW

File: src/msg/async/rdma/Infiniband.h [hunk 1]
```diff
@@ -241,34 +241,67 @@ class Infiniband {
...
-      static unsigned n_bufs_allocated;
-      static unsigned max_bufs;
-      static PerfCounters *perf_logger;
-     public:
-      typedef std::size_t size_type;
-      typedef std::ptrdiff_t difference_type;
-
-      static char * malloc(const size_type bytes);
-      static void free(char * const block);
+    class pool_context {
```
Comment (yuyuyu101) [inline src/msg/async/rdma/Infiniband.h]: MemPoolContext should be better

File: src/msg/async/rdma/Infiniband.cc [hunk 2]
```diff
@@ -633,45 +633,103 @@ int Infiniband::MemoryManager::Cluster::get_buffers(std::vector<Chunk*> &chunks,
...
+{
+  n_bufs_allocated += nbufs;
+
+  if (!perf_logger)
+    return;
 
+  if (nbufs > 0) {
+    perf_logger->inc(l_msgr_rdma_rx_bufs_total, nbufs);
+  } else {
+    perf_logger->dec(l_msgr_rdma_rx_bufs_total, -nbufs);
```
Comment (yuyuyu101) [inline src/msg/async/rdma/Infiniband.cc]: extra "-"

File: src/msg/async/rdma/Infiniband.cc [hunk 3]
```diff
@@ -633,45 +633,103 @@ int Infiniband::MemoryManager::Cluster::get_buffers(std::vector<Chunk*> &chunks,
...
+  }
+}
 
-char *Infiniband::MemoryManager::RxAllocator::malloc(const size_type bytes)
+void *Infiniband::MemoryManager::mem_pool::slow_malloc()
+{
+    void *p;
```
Comment (yuyuyu101) [inline src/msg/async/rdma/Infiniband.cc]: three space indent

File: src/msg/async/rdma/Infiniband.cc [hunk 4]
```diff
@@ -633,45 +633,103 @@ int Infiniband::MemoryManager::Cluster::get_buffers(std::vector<Chunk*> &chunks,
   return r;
 }
 
-Infiniband::MemoryManager* Infiniband::MemoryManager::RxAllocator::manager = nullptr;
-PerfCounters *Infiniband::MemoryManager::RxAllocator::perf_logger = nullptr;
+bool Infiniband::MemoryManager::pool_context::can_alloc(unsigned nbufs)
+{
+  /* unimited */
```
Comment (amitkumar50) [inline src/msg/async/rdma/Infiniband.cc]: @alex-mikheev Hope you mean unlimited

File: src/msg/async/rdma/Infiniband.h [hunk 5]
```diff
@@ -241,34 +241,67 @@ class Infiniband {
...
+        PerfCounters *perf_logger;
+
+      public:
+        MemoryManager *manager;
+        unsigned n_bufs_allocated;
```
Comment (amitkumar50) [inline src/msg/async/rdma/Infiniband.h]: @alex-mikheev Spacing looks not to be correct

File: src/msg/async/rdma/Infiniband.h [hunk 6]
```diff
@@ -241,34 +241,67 @@ class Infiniband {
...
+
+        static pool_context  *g_ctx;
+        static Mutex lock;
+    };
 
-      static void set_perf_logger(PerfCounters *logger) {
-        perf_logger = logger;
-        perf_logger->set(l_msgr_rdma_rx_bufs_total, n_bufs_allocated);
-      }
+    // modify boost pool so that it is possible to
```
Comment (amitkumar50) [inline src/msg/async/rdma/Infiniband.h]: @alex-mikheev I feel its always better to keep multi line comment in /* */

**Comment by alex-mikheev** [comment]
Thank you for the reviews. I will update the PR.

**Comment by yuyuyu101** [comment]
I found "Infiniband::post_chunks_to_srq" is too slow when init. each ceph command will stuck into Infiniband::post_chunks_to_srq nearly 5-8 seconds.

maybe we need to reserve a large memory range when init instead of malloc each

File: src/msg/async/rdma/Infiniband.cc (hunk 2 above)
Comment (alex-mikheev) [inline src/msg/async/rdma/Infiniband.cc]: I think dec() needs a positive value, thats why -nbufs

**Comment by alex-mikheev** [comment]
@yuyuyu101 actually there is only one allocation/registration of 2*ms_async_rdma_receive_queue_len and unless all receive buffers are used there should be no additional allocation. 

How I can reproduce the problem ? With ceph -s I am getting about 1 second overall run time in vstart environment.

> [alexm@andvari08 build]$ time bin/ceph -s
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2017-08-20 08:41:48.811515 7f938d5fa700 -1 WARNING: all dangerous and experimental features are enabled.
2017-08-20 08:41:48.858303 7f938d5fa700 -1 WARNING: all dangerous and experimental features are enabled.
  cluster:
    id:     bec5856d-2688-4c18-8982-6373b42ea237
    health: HEALTH_OK
>
> services:
    mon: 1 daemons, quorum a
    mgr: x(active)
    osd: 3 osds: 3 up, 3 in
>
>  data:
    pools:   1 pools, 8 pgs
    objects: 0 objects, 0 bytes
    usage:   202 GB used, 1447 GB / 1650 GB avail
    pgs:     8 active+clean
>
> real    0m1.058s
user    0m0.385s
sys     0m0.510s

File: src/msg/async/rdma/Infiniband.cc (hunk 2 above)
Comment (yuyuyu101) [inline src/msg/async/rdma/Infiniband.cc]: do you verify? I think not.... 

void PerfCounters::dec(int idx, uint64_t amt)
{
  if (!m_cct->_conf->perf)
    return;

  assert(idx > m_lower_bound);
  assert(idx < m_upper_bound);
  perf_counter_data_any_d& data(m_data[idx - m_lower_bound - 1]);
  assert(!(data.type & PERFCOUNTER_LONGRUNAVG));
  if (!(data.type & PERFCOUNTER_U64))
    return;
  data.u64 -= amt;
}

Comment (alex-mikheev) [inline src/msg/async/rdma/Infiniband.cc]: So dec() takes an unsigned value and decrements it from the counter. nbufs is negative and must be converted to positive. Thats why it is dec(...,-nbufs)

Comment (yuyuyu101) [inline src/msg/async/rdma/Infiniband.cc]: oh, sorry. my bad. ...

**Comment by yuyuyu101** [comment]
@alex-mikheev do you test under roce? I found memory manager assert when just boot

src/msg/async/rdma/Infiniband.cc: 945: FAILED assert(chunk != __null) (null)

**Comment by alex-mikheev** [comment]
@yuyuyu101 yes.  I check that i can run things under vstart environment. 
The commit itself is a cherry pick from the https://github.com/Mellanox/ceph/tree/luminous-rdma
which is tested on a cluster.

**Comment by yuyuyu101** [comment]
@alex-mikheev I can't run this pr. 

tcmalloc: large alloc 45176905728 bytes == 0x7fa6d69a4000 @  0x7fa6c8b31bf3 0x7fa6c8b53f23 0x7fa6cb9c0add 0x7fa6cb9c0e5f 0x7fa6cb9b93b8 0x7fa6cb9ba1ee 0x7fa6cb7e7652 0x7fa6cb7b448e 0x7fa6cb681bc3 0x7fa6cb681ff5 0x7fa6cb6829ea 0x7fa6cb5c2e12 0x7fa6c792cb35 0x7fa6cb651021 (nil)

**Comment by yuyuyu101** [comment]
you could check this pr with master

**Comment by alex-mikheev** [comment]
@yuyuyu101 sure we will check. Can you please paste an output of ibstat, ofed_info -s and free ?

**Comment by yuyuyu101** [comment]
MLNX_OFED_LINUX-4.1-1.0.2.0 (OFED-4.1-1.0.2):

ar_mgr:
osm_plugins/ar_mgr/ar_mgr-1.0-0.34.g9bd7c9a.tar.gz

cc_mgr:
osm_plugins/cc_mgr/cc_mgr-1.0-0.33.g9bd7c9a.tar.gz

dapl:
dapl.git mlnx_ofed_4_0
commit bdb055900059d1b8d5ee8cdfb457ca653eb9dd2d
dump_pr:
osm_plugins/dump_pr//dump_pr-1.0-0.29.g9bd7c9a.tar.gz

fabric-collector:
fabric_collector//fabric-collector-1.1.0.MLNX20170103.89bb2aa.tar.gz

hcoll:
mlnx_ofed_hcol/hcoll-3.8.1649-1.src.rpm

ibacm:
mlnx_ofed/ibacm.git mlnx_ofed_4_1
commit b0d53cf13358eb0c14665765b0170a37768463ff
ibacm_ssa:
mlnx_ofed_ssa/acm/ibacm_ssa-0.0.9.3.MLNX20151203.50eb579.tar.gz

ibdump:
sniffer/sniffer-5.0.0-1/ibdump/linux/ibdump-5.0.0-1.tgz

ibsim:
mlnx_ofed_ibsim/ibsim-0.6mlnx1-0.8.g9d76581.tar.gz

ibssa:
mlnx_ofed_ssa/distrib/ibssa-0.0.9.3.MLNX20151203.50eb579.tar.gz

ibutils:
ofed-1.5.3-rpms/ibutils/ibutils-1.5.7.1-0.12.gdcaeae2.tar.gz

ibutils2:
ibutils2/ibutils2-2.1.1-0.91.MLNX20170612.g2e0d52a.tar.gz

infiniband-diags:
mlnx_ofed_infiniband_diags/infiniband-diags-1.6.7.MLNX20170511.7595646.tar.gz

iser:
mlnx_ofed/mlnx-ofa_kernel-4.0.git mlnx_ofed_4_1
commit c22af8878c71966728f6ac38d963190f5222b2ec

isert:
mlnx_ofed/mlnx-ofa_kernel-4.0.git mlnx_ofed_4_1
commit c22af8878c71966728f6ac38d963190f5222b2ec

kernel-mft:
mlnx_ofed_mft/kernel-mft-4.7.0-41.src.rpm

knem:
knem.git mellanox-master
commit 4faa2978ad0339c50dd6df336d0a4182647b624b
libibcm:
mlnx_ofed/libibcm.git mlnx_ofed_4_1
commit e3e9fffe4d2d2f730110a7bdeb7da7b8ea97e51e
libibmad:
mlnx_ofed_libibmad/libibmad-1.3.13.MLNX20170511.267a441.tar.gz

libibprof:
mlnx_ofed_libibprof/libibprof-1.1.41-1.src.rpm

libibumad:
mlnx_ofed_libibumad/libibumad-13.10.2.MLNX20170511.dcc9f7a.tar.gz

libibverbs:
mlnx_ofed/libibverbs.git mlnx_ofed_4_1
commit a23bf787eff96af4c05d6e5f0e201dba80db114e
libmlx4:
mlnx_ofed/libmlx4.git mlnx_ofed_4_1
commit d945a7eeb52e319b2199990e6602a9fee0646371
libmlx5:
mlnx_ofed/libmlx5.git mlnx_ofed_4_1
commit 71822e375014c7f81dec3e4eca06f366846eaf1a
libopensmssa:
mlnx_ofed_ssa/plugin/libopensmssa-0.0.9.3.MLNX20151203.50eb579.tar.gz

librdmacm:
mlnx_ofed/librdmacm.git mlnx_ofed_4_1
commit 1297178df9b07030d84a042d417cb61fa65e62a1
librxe:
mlnx_ofed/librxe.git master
commit 607460456c717c3b65428367676cacb5495ac005
libvma:
vma/source_rpms//libvma-8.3.7-0.src.rpm

mlnx-en:
mlnx_ofed/mlnx-ofa_kernel-4.0.git mlnx_ofed_4_1
commit c22af8878c71966728f6ac38d963190f5222b2ec

mlnx-ethtool:
upstream/ethtool.git for-upstream
commit ac0cf295abe0c0832f0711fed66ab9601c8b2513
mlnx-nfsrdma:
mlnx_ofed/mlnx-ofa_kernel-4.0.git mlnx_ofed_4_1
commit c22af8878c71966728f6ac38d963190f5222b2ec

mlnx-nvme:
mlnx_ofed/mlnx-ofa_kernel-4.0.git mlnx_ofed_4_1
commit c22af8878c71966728f6ac38d963190f5222b2ec

mlnx-ofa_kernel:
mlnx_ofed/mlnx-ofa_kernel-4.0.git mlnx_ofed_4_1
commit c22af8878c71966728f6ac38d963190f5222b2ec

mlnx-rdma-rxe:
mlnx_ofed/mlnx-ofa_kernel-4.0.git mlnx_ofed_4_1
commit c22af8878c71966728f6ac38d963190f5222b2ec

mpi-selector:
ofed-1.5.3-rpms/mpi-selector/mpi-selector-1.0.3-1.src.rpm

mpitests:
mlnx_ofed_mpitest/mpitests-3.2.19-acade41.src.rpm

mstflint:
mlnx_ofed_mstflint/mstflint-4.7.0-1.6.g26037b7.tar.gz

multiperf:
mlnx_ofed_multiperf/multiperf-3.0-0.10.gda89e8c.tar.gz

mxm:
mlnx_ofed_mxm/mxm-3.6.3102-1.src.rpm

ofed-docs:
docs.git mlnx_ofed-4.0
commit 3d1b0afb7bc190ae5f362223043f76b2b45971cc

openmpi:
mlnx_ofed_ompi_1.8/openmpi-2.1.2a1-1.src.rpm

opensm:
mlnx_ofed_opensm/opensm-4.9.0.MLNX20170607.280b8f7.tar.gz

perftest:
mlnx_ofed_perftest/perftest-4.1-0.4.g16dbf63.tar.gz

qperf:
mlnx_ofed_qperf/qperf-0.4.9.tar.gz

sharp:
mlnx_ofed_sharp/sharp-1.3.1.MLNX20170625.859dc24.tar.gz

sockperf:
sockperf/sockperf-3.1-14.gita9f6056282ef.src.rpm

srp:
mlnx_ofed/mlnx-ofa_kernel-4.0.git mlnx_ofed_4_1
commit c22af8878c71966728f6ac38d963190f5222b2ec

srptools:
srptools/srptools-41mlnx1-4.src.rpm

ucx:
mlnx_ofed_ucx/ucx-1.2.2947-1.src.rpm


Installed Packages:
-------------------

kmod-srp
libibumad
dapl-devel
mxm
libibmad
perftest
hcoll
kmod-mlnx-ofa_kernel
libmlx5
librdmacm-devel
dump_pr
knem-mlnx
librxe-devel-static
opensm-devel
infiniband-diags
kmod-kernel-mft-mlnx
libibverbs
librxe
libibmad-devel
opensm
mstflint
ibdump
openmpi
kmod-iser
libibverbs-utils
libibcm-devel
ibacm
dapl
ibutils2
qperf
mlnxofed-docs
kmod-knem-mlnx
libibverbs-devel-static
libibcm
ibsim
opensm-static
srptools
infiniband-diags-compat
mlnx-ethtool
mlnx-ofa_kernel
kmod-isert
libmlx4-devel
libibumad-devel
librdmacm-utils
dapl-devel-static
cc_mgr
ucx
libmlx4
librdmacm
ibutils
mpitests_openmpi
mlnx-ofa_kernel-devel
libmlx5-devel
opensm-libs
ar_mgr
mpi-selector
libibumad-static
dapl-utils
sharp
libibverbs-devel
libibmad-static
libibprof

**Comment by yuyuyu101** [comment]
# ibstat
CA 'mlx4_0'
        CA type: MT4099
        Number of ports: 2
        Firmware version: 2.40.7000
        Hardware version: 1
        Node GUID: 0x7cfe900300917c00
        System image GUID: 0x7cfe900300917c03
        Port 1:
                State: Active
                Physical state: LinkUp
                Rate: 56
                Base lid: 1
                LMC: 0
                SM lid: 1
                Capability mask: 0x02514868
                Port GUID: 0x7cfe900300917c01
                Link layer: InfiniBand
        Port 2:
                State: Active
                Physical state: LinkUp
                Rate: 56
                Base lid: 2
                LMC: 0
                SM lid: 1
                Capability mask: 0x02514868
                Port GUID: 0x7cfe900300917c02
                Link layer: InfiniBand

File: src/msg/async/rdma/Infiniband.cc [hunk 7]
```diff
@@ -633,45 +633,103 @@ int Infiniband::MemoryManager::Cluster::get_buffers(std::vector<Chunk*> &chunks,
...
+        " limit: " << manager->cct->_conf->ms_async_rdma_receive_buffers << dendl;
+    return false;
+  }
+
+  return true;
+}
+
+void Infiniband::MemoryManager::MemPoolContext::set_stat_logger(PerfCounters *logger) {
+  perf_logger = logger;
+  if (perf_logger != nullptr)
```
Comment (amitkumar50) [inline src/msg/async/rdma/Infiniband.cc]: @alex-mikheev wouldn't be good to check nullptr 1st then assignment?
if (logger !=  nullptr)
     perf_logger = logger;
     perf_logger->set(l_msgr_rdma_rx_bufs_total, n_bufs_allocated);

**Pull request merged**